File: scripts/pid_reload.py
#!/usr/bin/env python3
import rospy
from std_msgs.msg import Float32
from geometry_msgs.msg import PoseStamped
import math
import time

POSE_TOPIC = "vrpn_client_node/JaiAliJetRacer/pose"
ERR_EPSILON = 0.1
GOAL = [-3.0, 3.0]

class PIDController:
    def __init__(self, kp, ki, kd):
        self.kp, self.ki, self.kd = kp, ki, kd
        self.integral = 0
        self.last_error = 0
        self.last_time = time.time()

    def reset(self):
        self.integral = 0
        self.last_error = 0
        self.last_time = time.time()

    def update(self, error):
        current_time = time.time()
        dt = current_time - self.last_time
        de = error - self.last_error

        self.integral += error * dt
        derivative = de / dt if dt > 0 else 0

        output = self.kp * error + self.ki * self.integral + self.kd * derivative

        self.last_error = error
        self.last_time = current_time

        return output

class JetRacerController:
    def __init__(self):
        rospy.init_node('jetracer_pid_controller')
        self.steering_pub = rospy.Publisher('/jetracer/steering', Float32, queue_size=1)
        self.throttle_pub = rospy.Publisher('/jetracer/throttle', Float32, queue_size=1)
        rospy.Subscriber(POSE_TOPIC, PoseStamped, self.pose_callback)

        self.goal = GOAL  # example goal point in meters

        self.heading_pid = PIDController(1.0, 0.0, 0.65)
        self.distance_pid = PIDController(0.5, 0.0, 0.70) # Try pure proportinality first

    def pose_callback(self, msg):
        x = msg.pose.position.x
        y = msg.pose.position.y
        print('Postion: ', x, y)
        q = msg.pose.orientation
        _, _, yaw = euler_from_quaternion(q.x, q.y, q.z, q.w)

        dx = self.goal[0] - x
        dy = self.goal[1] - y
        goal_dist = math.hypot(dx, dy)
        goal_heading = math.atan2(dy, dx)
        heading_error = self.angle_wrap(goal_heading - yaw)

        # Compute control
        steering = self.heading_pid.update(heading_error)
        throttle = self.distance_pid.update(goal_dist)

        # Limit values
        steering = max(min(steering, 1.0), -1.0)
        throttle = max(min(throttle, 0.2), 0.0) if goal_dist > ERR_EPSILON else 0.0

        self.steering_pub.publish(Float32(steering))
        self.throttle_pub.publish(Float32(throttle))

    def angle_wrap(self, angle):
        while angle > math.pi:
            angle -= 2 * math.pi
        while angle < -math.pi:
            angle += 2 * math.pi
        return angle

def euler_from_quaternion(x, y, z, w):
        """
        Convert a quaternion into euler angles (roll, pitch, yaw)
        roll is rotation around x in radians (counterclockwise)
        pitch is rotation around y in radians (counterclockwise)
        yaw is rotation around z in radians (counterclockwise)
        """
        t0 = +2.0 * (w * x + y * z)
        t1 = +1.0 - 2.0 * (x * x + y * y)
        roll_x = math.atan2(t0, t1)
     
        t2 = +2.0 * (w * y - z * x)
        t2 = +1.0 if t2 > +1.0 else t2
        t2 = -1.0 if t2 < -1.0 else t2
        pitch_y = math.asin(t2)
     
        t3 = +2.0 * (w * z + x * y)
        t4 = +1.0 - 2.0 * (y * y + z * z)
        yaw_z = math.atan2(t3, t4)
     
        return roll_x, pitch_y, yaw_z # in radians


if __name__ == '__main__':
    controller = JetRacerController()
    rospy.spin()

File: scripts/pid_reload.jl
#!/usr/bin/env julia

#=
Authors: Jai Nagaraj, Ali Chimoun
The purpose of this script is to provide a boilerplate for PID control on NVIDIA JetRacers in Julia.
This is essentially a translated version of the existing Python script using the RobotOS.jl package.
=#

using RobotOS
@rosimport std_msgs.msg: Float32
@rosimport geometry_msgs.msg: PoseStamped
rostypegen()

using .std_msgs.msg
using .geometry_msgs.msg
using Dates
using LinearAlgebra

const POSE_TOPIC = "vrpn_client_node/JaiAliJetRacer/pose"
const ERR_EPSILON = 0.1
const GOAL = [-3.0, -3.0]

#=
**** PIDCONTROLLER CLASS STUFF ****
Since Julia does not have full OOP, code is manually organized into classes.
=#

# Julia version of the Python PIDController class.
mutable struct PIDController
    kp::Float64
    ki::Float64
    kd::Float64
    integral::Float64
    last_error::Float64
    last_time::DateTime
end

function PIDController(kp, ki, kd)
    now = nowtime()
    PIDController(kp, ki, kd, 0.0, 0.0, now)
end

function reset!(pid::PIDController)
    pid.integral = 0.0
    pid.last_error = 0.0
    pid.last_time = nowtime()
end

function update!(pid::PIDController, error::Float64)
    current_time = nowtime()
    dt = max(1e-6, Dates.value(current_time - pid.last_time) / 1e3)  # seconds
    de = error - pid.last_error

    pid.integral += error * dt
    derivative = de / dt

    output = pid.kp * error + pid.ki * pid.integral + pid.kd * derivative

    pid.last_error = error
    pid.last_time = current_time

    return output
end

#=
**** JETRACERCONTROLLER CLASS STUFF ****
Since Julia does not have full OOP, code is manually organized into classes.
=#

struct JetRacerController
    steering_pub # Publisher to jetracer steering
    throttle_pub # Publisher to jetracer throttle
    heading_pid::PIDController # controls values published to steering
    distance_pid::PIDController # controls values published to throttle
    goal::Vector{Float64} # may change depending on trajectory/purpose
end

function JetRacerController()
    init_node("jetracer_pid_controller")
    steering_pub = Publisher{Float32}("/jetracer/steering", queue_size=1)
    throttle_pub = Publisher{Float32}("/jetracer/throttle", queue_size=1)
    controller = JetRacerController(
        steering_pub,
        throttle_pub,
        PIDController(1.0, 0.0, 0.65),
        PIDController(0.5, 0.0, 0.70),
        GOAL
    )
    Subscriber(POSE_TOPIC, PoseStamped, msg -> pose_callback(controller, msg))
    return controller
end

function pose_callback(controller::JetRacerController, msg::PoseStamped)
    x = msg.pose.position.x
    y = msg.pose.position.y
    println("Position: $x, $y")
    q = msg.pose.orientation
    _, _, yaw = euler_from_quaternion(q.x, q.y, q.z, q.w)

    dx = controller.goal[1] - x
    dy = controller.goal[2] - y
    goal_dist = hypot(dx, dy)
    goal_heading = atan(dy, dx)
    heading_error = angle_wrap(goal_heading - yaw)

    steering = update!(controller.heading_pid, heading_error)
    throttle = update!(controller.distance_pid, goal_dist)

    steering = clamp(steering, -1.0, 1.0)
    throttle = goal_dist > ERR_EPSILON ? clamp(throttle, 0.0, 0.2) : 0.0

    publish(controller.steering_pub, Float32(data=steering))
    publish(controller.throttle_pub, Float32(data=throttle))
end


#=
**** UTILITY FUNCTIONS ****
=#

function angle_wrap(angle::Float64)
    while angle > π
        angle -= 2π
    end
    while angle < -π
        angle += 2π
    end
    return angle
end

function euler_from_quaternion(x, y, z, w)
    t0 = +2.0 * (w * x + y * z)
    t1 = +1.0 - 2.0 * (x * x + y * y)
    roll_x = atan(t0, t1)

    t2 = +2.0 * (w * y - z * x)
    t2 = clamp(t2, -1.0, 1.0)
    pitch_y = asin(t2)

    t3 = +2.0 * (w * z + x * y)
    t4 = +1.0 - 2.0 * (y * y + z * z)
    yaw_z = atan(t3, t4)

    return roll_x, pitch_y, yaw_z
end

function nowtime()
    return now(UTC)
end

# --- Main ---
if abspath(PROGRAM_FILE) == @__FILE__
    controller = JetRacerController()
    spin()
end
